core/domain/classifier_services.py

```diff
@@ -0,0 +1,338 @@
+        """Returns a label's id if it exists, otherwise assigns
+        a new id to the label and returns it."""
+        if label not in self._label_to_id:
+            self._label_to_id[label] = self._num_labels
+            self._num_labels += 1
+        return self._label_to_id[label]
+
+    def _get_record(self, d):
+        """Given a doc id, return the doc and labels."""
+        return self._w_dp[d], self._b_dl[d]
+
+    def _get_label_vector(self, labels):
+        """Returns a vector specifying which labels
+        are specified in the training set."""
+        # When we see an empty label list, predicting on all labels
+        # is assumed (versus having to pass all the labels each prediction)
+        if len(labels) == 0:
+            return numpy.ones(self._num_labels)
+        label_vector = numpy.zeros(self._num_labels)
+        for label in labels:
+            label_vector[self._get_label_id(label)] = 1
+        # Always set default label
+        label_vector[self._label_to_id[self._DEFAULT_LABEL]] = 1
+        return label_vector
+
+    def _update_counting_matrices(self, d, w, l, inc=True):
+        """Update counting matrices given doc/word/label indices.
+        Either increments or decrements."""
+        val = 1 if inc else -1
+        self._c_dl[d, l] += val
+        self._c_lw[l, w] += val
+        self._c_l[l] += val
+
+    def _get_doc_ids(self, doc_ids):
+        if doc_ids is None:
+            doc_ids = xrange(self._num_docs)
+        return doc_ids
+
+    def _init_docs(self, doc_ids=None):
+        """Initialize data for given docs (defaults to all)."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            l_c = [
+                numpy.random.multinomial(1, labels / labels.sum()).argmax()
+                for i in xrange(len(doc))
+            ]
+            self._l_dp.append(l_c)
+            for w, l in zip(doc, l_c):
+                self._update_counting_matrices(d, w, l)
+
+    def _infer_docs(self, doc_ids):
+        """Runs iterative inference on given docs."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        statez = {
+            'updates': 0,
+            'computes': 0
+        }
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            for p, w in enumerate(doc):
+                l = self._l_dp[d][p]
+
+                self._update_counting_matrices(d, w, l, False)
+
+                # Gibbs update of labels
+                coeff_a = 1.0 / (
+                    self._c_dl[d].sum() + self._num_labels * self._alpha)
+                coeff_b = 1.0 / (
+                    self._c_lw.sum(axis=1) + self._num_words * self._beta)
+                prob_l = (
+                    labels *
+                    coeff_a * (self._c_dl[d] + self._alpha) *
+                    coeff_b * (self._c_lw[:, w] + self._beta))
+                new_l = numpy.random.multinomial(
```

Comment:
I'm uncomfortable seeing both prob_l and new_l here, since you're using *_l to mean different things in each context. I think maybe the best thing to do is to use label_id and new_label_id instead (and keep prob_l as-is).